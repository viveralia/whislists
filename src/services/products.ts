import firebase from "firebase/app";

import { serverTimestamp } from "../firebase";
import { getScreenshot } from "./screenshot";
import { wishListsRef, WishList } from "./wishlists";

export interface Product {
  id: string;
  name: string;
  url: string;
  imageUrl: string;
  createdBy: {
    name: string;
    id: string;
  };
}

export type NewProduct = Pick<Product, "name" | "url">;

export async function addProductToWishList(
  wishList: WishList,
  user: firebase.User,
  newProduct: NewProduct
) {
  return await wishListsRef
    .doc(wishList.id)
    .collection("products")
    .add({
      ...newProduct,
      imageUrl: await getScreenshot(newProduct.url),
      createdBy: {
        name: user.displayName,
        id: user.uid,
      },
      createdAt: serverTimestamp(),
    });
}

export function suscribeToProductsInAWishList(
  wishList: WishList,
  saveCallback: (wishLists: Product[]) => void
) {
  return wishListsRef
    .doc(wishList.id)
    .collection("products")
    .onSnapshot((snap) => {
      const products = snap.docs.map((doc) => ({
        id: doc.id,
        ...(doc.data() as Omit<Product, "id">),
      }));
      saveCallback(products);
    });
}

export function removeProductFromAWishList(
  wishList: WishList,
  product: Product
) {
  return wishListsRef
    .doc(wishList.id)
    .collection("products")
    .doc(product.id)
    .delete();
}

export async function updateProductInAWishlist(
  wishList: WishList,
  product: Product
) {
  const productRef = wishListsRef
    .doc(wishList.id)
    .collection("products")
    .doc(product.id);

  const doc = await productRef.get();

  if (!doc.exists) throw Error("Can't update a non existing product");

  const isUrlTheSame = doc.data()?.url === product.url;

  return await productRef.update({
    name: product.name,
    url: product.url,
    imageUrl: isUrlTheSame
      ? product.imageUrl
      : await getScreenshot(product.url),
  });
}
