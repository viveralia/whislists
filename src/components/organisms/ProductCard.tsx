import { FC } from "react";

import { Product } from "../../services/products";

export interface ProductCardProps {
  product: Product;
  onDelete: () => Promise<void> | void;
  onEdit: () => void;
}

const ProductCard: FC<ProductCardProps> = ({ product, onDelete, onEdit }) => {
  return (
    <article>
      <h3>{product.name}</h3>
      <button onClick={onEdit}>Edit</button>
      <button onClick={onDelete}>Remove</button>
      <a
        className="block"
        href={product.url}
        target="_blank"
        rel="noopener noreferrer"
      >
        <img src={product.imageUrl} alt={product.name} />
      </a>
    </article>
  );
};

export default ProductCard;
