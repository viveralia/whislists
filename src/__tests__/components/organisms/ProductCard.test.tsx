import { shallow } from "enzyme";

import ProductCard from "../../../components/organisms/ProductCard";
import products from "../../../fixtures/products";

it("should render correctly", () => {
  const wrapper = shallow(
    <ProductCard
      product={products[0]}
      onDelete={() => console.log("Deleting")}
      onEdit={() => console.log("Editing")}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
