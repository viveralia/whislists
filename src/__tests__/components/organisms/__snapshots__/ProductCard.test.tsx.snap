// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<article>
  <h3>
    Product 1
  </h3>
  <button
    onClick={[Function]}
  >
    Edit
  </button>
  <button
    onClick={[Function]}
  >
    Remove
  </button>
  <a
    className="block"
    href="https://google.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      alt="Product 1"
      src="https://google.com"
    />
  </a>
</article>
`;
